### PR TITLE
feat(indev): add lv_indev_set_mode to switch input device's event model

### PR DESCRIPTION
**Pull request by bjsylvia**
Reference Implementation for https://github.com/lvgl/lvgl/issues/4945

### Description of the feature or fix

A clear and concise description of what the bug or new feature is.

### Checkpoints
- [ ] Run `code-format.py` from the scripts folder. [astyle](http://astyle.sourceforge.net/install.html) needs to be installed.
- [ ] Update the [Documentation](https://github.com/lvgl/lvgl/tree/master/docs) if needed
- [ ] Add [Examples](https://github.com/lvgl/lvgl/tree/master/examples) if relevant. 
- [ ] Add [Tests](https://github.com/lvgl/lvgl/blob/master/tests/README.md) if applicable.
- [ ] If you added new options to `lv_conf_template.h` run [lv_conf_internal_gen.py](https://github.com/lvgl/lvgl/blob/release/v8.3/scripts/lv_conf_internal_gen.py) and update [Kconfig](https://github.com/lvgl/lvgl/blob/release/v8.3/Kconfig).
 
Be sure the following conventions are followed:
- [ ] Follow the [Styling guide](https://github.com/lvgl/lvgl/blob/master/docs/CODING_STYLE.md)
- [ ] Prefer `enum`s instead of macros. If inevitable to use `define`s export them with `LV_EXPORT_CONST_INT(defined_value)` right after the `define`.
- [ ] In function arguments prefer `type name[]` declaration for array parameters instead of `type * name`
- [ ] Use typed pointers instead of `void *` pointers
- [ ] Do not `malloc` into a static or global variables. Instead declare the variable in `lv_global_t` structure in [`lv_global.h`](https://github.com/lvgl/lvgl/blob/master/src/core/lv_global.h) and mark the variable with `(LV_GLOBAL_DEFAULT()->variable)` when it's used. See a detailed description [here](https://docs.lvgl.io/master/get-started/bindings/micropython.html#memory-management).
- [ ] Widget constructor must follow the `lv_<widget_name>_create(lv_obj_t * parent)` pattern.
- [ ] Widget members function must start with `lv_<module_name>` and should receive `lv_obj_t *` as first argument which is a pointer to widget object itself.  
- [ ] `struct`s should be used via an API and not modified directly via their elements.
- [ ] `struct` APIs should follow the widgets' conventions. That is to receive a pointer to the `struct` as the first argument, and the prefix of the `struct` name should be used as the prefix of the function name too (e.g.  `lv_disp_set_default(lv_disp_t * disp)`)
- [ ] Functions and `struct`s which are not part of the public API must begin with underscore in order to mark them as "private".
- [ ] Arguments must be named in H files too.
- [ ] To register and use callbacks one of the following needs to be followed (see a detailed description [here](https://docs.lvgl.io/master/get-started/bindings/micropython.html#callbacks)):
  - For both the registration function and the callback pass a pointer to a `struct` as the first argument. The `struct` must contain `void * user_data` field.
  - The last argument of the registration function must be `void * user_data` and the same `user_data` needs to be passed as the last argument of the callback.
  - Callback types not following these conventions should end with `xcb_t`.

### CODE REVIEW

**Comment by bjsylvia** [comment]
@kisvegabor This PR only processes Pointer device. Please check the feasibility of this plan and whether there are any uncovered scenarios.

We can add Keypad and Encoder device later.

**Comment by XuNeo** [comment]
Another similar requirement is to detect double-click and triple-click for touch, button etc. input device. I have written a routine to deal with common key event detection. 

It would be nice integrate to `lvgl`'s `indev` processing.

```c

static struct btn_state_st {
    uint8_t btn; /* self */
    struct {
        bool state; /* pressed or not*/
        uint32_t timestamp; /* timestamp of when this state updated */
    } states[STATE_ARRAY_LEN]; /* 6 state to detect triple-click */
    uint8_t windex; /* write index to state array */

    bool countingdown;
    uv_timer_t timer; /* on-shot timer for longpress detection */
    uint8_t lp_index; /* the index when timer started for longpress */
} g_btn[LVX_BTN_COUNT];

static void enqueue_event(int which, int event, int pressed_time){ }

static uint32_t tick_diff(uint32_t now_tick, uint32_t prev_tick)
{
    if (now_tick >= prev_tick) prev_tick = now_tick - prev_tick;
    else {
        prev_tick = UINT32_MAX - prev_tick + 1;
        prev_tick += now_tick;
    }
    return prev_tick;
}

static void longpress_timer_cb(uv_timer_t* handle)
{
    struct btn_state_st* btn = handle->data;
    uint32_t elapsed = lv_tick_elaps(btn->states[btn->lp_index].timestamp);

    if (elapsed < LONGPRESS_THRESHOLD)
        return;
    enqueue_event(btn->btn, LVX_INPUT_EVENT_LONG_PRESSED, elapsed);
}

/* button state changed, get event based on state */
static void inline btn_event_parse(int which)
{
    uint32_t time_diff;
    struct btn_state_st* btn = &g_btn[which];
    int now_index = btn->windex;
    int pre_index = PREV(now_index);

    bool pressed = btn->states[now_index].state;

    if (pressed) {
        /* start timer to check long press */
        enqueue_event(which, LVX_INPUT_EVENT_PRESSED, 0);
        btn->countingdown = true;
        btn->lp_index = now_index;
        uv_update_time(g_loop);
        uv_timer_start(&btn->timer, longpress_timer_cb, LONGPRESS_CHECK_PERIOD,
                       LONGPRESS_CHECK_PERIOD);
    } else {
        /* released, stop timer firstly */
        if (btn->countingdown) {
            btn->countingdown = false;
            uv_timer_stop(&btn->timer);
        }

        /* send out released event  */
        enqueue_event(which, LVX_INPUT_EVENT_RELEASED, 0);

        /* short click */
        if (btn->longpress_state) {
            /* if longpress_state, no double/triple click check */
            btn->longpress_state = 0; /* 0 means no long press */
            return;
        } else {
            enqueue_event(which, LVX_INPUT_EVENT_SHORT_CLICKED, 0);
        }

        /* triple click detect */
        pre_index = NEXT(now_index);

        time_diff = tick_diff(btn->states[now_index].timestamp,
                              btn->states[pre_index].timestamp);
        if (time_diff < TRIPLE_CLICK_TIMEOUT) {
            /* triple click in 0.9 seconds */
            enqueue_event(which, LVX_INPUT_EVENT_TRIPLE_CLICKED, 0);
            /* clear the state */
            for (int i = 0; i < STATE_ARRAY_LEN; i++) {
                if (i == now_index)
                    continue;
                btn->states[i].state = false;
                btn->states[i].timestamp = 0;
            }

            /* ignore the double click event */
            return;
        }

        /* double click detect */
        pre_index = (now_index - 3 + STATE_ARRAY_LEN) % STATE_ARRAY_LEN;
        time_diff = tick_diff(btn->states[now_index].timestamp,
                              btn->states[pre_index].timestamp);
        if (time_diff < DOUBLE_CLICK_TIMEOUT) {
            /* double click in xx seconds */
            enqueue_event(which, LVX_INPUT_EVENT_DOUBLE_CLICKED, 0);
        }
    }
}

/*Called from timer or from event callback*/
static void inline btn_state_update(int which, bool state)
{
    struct btn_state_st* btn = &g_btn[which];
    int windex = btn->windex;
    uint32_t timestamp = lv_tick_get();
    btn->states[windex].state = state;
    btn->states[windex].timestamp = timestamp;
    btn_event_parse(which);

    /* increase the write handler after state parser */
    btn->windex = NEXT(windex);
}


```

**Comment by kisvegabor** [comment]
@bjsylvia 
It could work, but I can't see the benefit compared to pausing the read_timer on release. Actually LVGL could automatically pause the timer when the release event is processed.

```c
lv_indev_set_auto_pause_on_release(indev, true/false); //Needs a better name

//On touch press event or interrupt
lv_indev_resume(indev);
```

It seems less invasive to me.

@XuNeo 
I absolutely like the idea adding double and triple click events but let's do it after v9.0 in January or February.

File: src/indev/lv_indev.c [hunk 1]
```diff
@@ -486,11 +487,23 @@ lv_timer_t * lv_indev_get_read_timer(lv_indev_t * indev)
...
-        lv_timer_delete(indev->read_timer);
-        indev->read_timer = NULL;
+    if(!indev || indev->mode == mode)
+        return;
+
+    indev->mode = mode;
+    if(indev->read_timer) {
+        if(mode == LV_INDEV_MODE_EVENT) {
+            lv_timer_pause(indev->read_timer);
+            lv_timer_set_period(indev->read_timer, LV_INDEV_DEF_LONG_PRESS_REP_TIME);
```
Comment (kisvegabor) [inline src/indev/lv_indev.c]:
```suggestion
            lv_timer_set_period(indev->read_timer, LV_DEF_REFR_PERIOD);
```

**Review by kisvegabor — COMMENTED** [review]
Looks good, just a minor comment.

File: src/indev/lv_indev.c (hunk 1 above)
Comment (bjsylvia) [reply src/indev/lv_indev.c]: If we keep the default refresh rate, maybe we should remove `lv_timer_set_period`?

Comment (kisvegabor) [reply src/indev/lv_indev.c]: I agree to remove this line.

**Comment by bjsylvia** [comment]
@kisvegabor I've updated it. Please review it.

**Comment by kisvegabor** [comment]
Looks good, and works well!

**Comment by kisvegabor** [comment]
Ah, I've just forgot the docs. Could write a line abut it [here](https://docs.lvgl.io/master/porting/indev.html#other-features)?

**Pull request merged**